qiskit/transpiler/passes/layout/sabre_pre_layout.py

```diff
@@ -0,0 +1,217 @@
+            else:
+                self.property_set["sabre_starting_layouts"].append(starting_layout)
+
+    def _add_extra_edges(self, distance):
+        """Augments the coupling map with extra edges that connect nodes ``distance``
+        apart in the original graph. The extra edges are assigned errors allowing VF2
+        to prioritize real edges over extra edges.
+        """
+        nq = len(self.coupling_map.graph.node_indices())
+        augmented_coupling_map = CouplingMap()
+        augmented_coupling_map.graph = self.coupling_map.graph.copy()
+        augmented_error_map = ErrorMap(nq)
+
+        for (x, y) in itertools.combinations(self.coupling_map.graph.node_indices(), 2):
+            d = self.coupling_map.distance(x, y)
+            if 1 < d <= distance:
+                augmented_coupling_map.add_edge(x, y)
+                augmented_error_map.add_error((x, y), self.error_rate)
+                augmented_coupling_map.add_edge(y, x)
+                augmented_error_map.add_error((y, x), self.error_rate)
+
+        return augmented_coupling_map, augmented_error_map
+
+    def _get_extra_edges_used(self, dag, layout):
+        """Returns the list of extra edges involved in the layout."""
+        extra_edges_used = []
+        virtual_bits = layout.get_virtual_bits()
+        for node in dag.op_nodes():
+            if len(node.qargs) == 2:
+                p0 = virtual_bits[node.qargs[0]]
+                p1 = virtual_bits[node.qargs[1]]
+                if self.coupling_map.distance(p0, p1) > 1:
+                    extra_edge = (p0, p1) if p0 < p1 else (p1, p0)
+                    extra_edges_used.append(extra_edge)
+        return extra_edges_used
+
+    def _find_layout(self, dag, edges):
+        """Checks if there is a layout for a given set of edges."""
+        cm = CouplingMap(edges)
+        pass_ = VF2Layout(cm, seed=0, max_trials=1, call_limit=self.call_limit_vf2)
+        pass_.run(dag)
+        return pass_.property_set.get("layout", None)
+
+    def _minimize_extra_edges(self, dag, starting_layout):
+        """Minimizes the set of extra edges involved in the layout. This iteratively
+        removes extra edges from the coupling map and uses VF2 to check if a layout
+        still exists. This is reasonably efficiently as it only looks for a local
+        minimum.
+        """
+        # compute the set of edges in the original coupling map
+        real_edges = []
+        for (x, y) in itertools.combinations(self.coupling_map.graph.node_indices(), 2):
+            d = self.coupling_map.distance(x, y)
+            if d == 1:
+                real_edges.append((x, y))
+
+        best_layout = starting_layout
+
+        # keeps the set of "necessary" extra edges: without a necessary edge
+        # a layout no longer exists
+        extra_edges_necessary = []
+
+        extra_edges = self._get_extra_edges_used(dag, starting_layout)
+        extra_edges_unprocessed_set = set(extra_edges)
```

Comment:
Would it be easier to just have `_get_extra_edges_used` internally collect edges into a set directly. I don't think this list will ever get super long, but we do have to iterate over it to create a set from the list.

Reply:
Done in [7235ef7](https://github.com/Qiskit/qiskit/pull/10829/commits/7235ef7b6b5c65db86ce4a51e891adcb042ccea5)